### PR TITLE
external validation and refactoring

### PR DESCRIPTION
**Pull request by kmoore-intuit**
# What Changed

# Why

Todo:

- [ ] Add tests
- [ ] Add docs

### CODE REVIEW

File: src/main/java/com/intuit/graphql/orchestrator/schema/transform/FederationTransformerPostMerge.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+  private TypeDefinition getBaseEntity(XtextGraph xtextGraph, String entityTypename, String serviceNamespace) {
+        TypeDefinition entityBaseType = xtextGraph.getEntitiesByTypeName().get(entityTypename);
+        if (Objects.isNull(entityBaseType)) {
+            throw new BaseTypeNotFoundException(entityTypename, serviceNamespace);
+        }
+        return entityBaseType;
+    }
+
+  private void validateBaseExtensionCompatibility(EntityMergingContext entityMergingContext) {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/schema/transform/FederationTransformerPostMerge.java:93]: 👍

File: src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java [hunk 2]
```diff
@@ -114,7 +114,7 @@ private BatchFieldAuthorization getDefaultOrCustomFieldAuthorization(GraphQLCont
       MergedField filteredRootField = result.getMergedField();
       if (filteredRootField != null) {
         filteredRootField.getFields().stream()
-            .map(field -> (serviceMetadata.hasFieldResolverDirective() || serviceMetadata.isFederationService())
+            .map(field -> (serviceMetadata.shouldRemoveExternalFields())
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java:117]: 👍

File: src/main/java/com/intuit/graphql/orchestrator/federation/ExternalValidator.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import com.intuit.graphql.orchestrator.federation.exceptions.ExternalFieldNotFoundInBaseException;
+import com.intuit.graphql.orchestrator.stitching.StitchingException;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.intuit.graphql.orchestrator.utils.XtextTypeUtils.getFieldDefinitions;
+
+public class ExternalValidator {
+    public void validatePostMerge(EntityTypeMerger.EntityMergingContext entityMergingContext, FieldDefinition fieldDefinition, Directive directive) {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/federation/ExternalValidator.java:14]: not sure if I missed it, looks like directive is unused?

File: src/main/java/com/intuit/graphql/orchestrator/federation/keydirective/KeyDirectiveValidator.java [hunk 4]
```diff
@@ -1,24 +1,28 @@
...
+import static com.intuit.graphql.orchestrator.utils.FederationConstants.FEDERATION_KEY_DIRECTIVE;
+import static com.intuit.graphql.orchestrator.utils.XtextUtils.getDirectivesWithNameFromDefinition;
+import static java.lang.String.format;
 
-/**
- * This class helps break up the {@link com.intuit.graphql.orchestrator.schema.transform.KeyTransformer} by
- * validating if the TypeDefinition key directive is valid.
- */
 public class KeyDirectiveValidator {
+  FieldSetValidator fieldSetValidator = new FieldSetValidator();
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/federation/keydirective/KeyDirectiveValidator.java:25]: final keyword

File: src/main/java/com/intuit/graphql/orchestrator/federation/requiresdirective/RequireValidator.java [hunk 5]
```diff
@@ -12,11 +13,11 @@
...
 import java.util.Optional;
 
-import static com.intuit.graphql.orchestrator.utils.FederationUtils.FEDERATION_FIELDS_ARGUMENT;
-import static com.intuit.graphql.orchestrator.utils.FederationUtils.FEDERATION_REQUIRES_DIRECTIVE;
-import static com.intuit.graphql.orchestrator.utils.FederationUtils.checkFieldSetValidity;
+import static com.intuit.graphql.orchestrator.utils.FederationConstants.FEDERATION_FIELDS_ARGUMENT;
+import static com.intuit.graphql.orchestrator.utils.FederationConstants.FEDERATION_REQUIRES_DIRECTIVE;
 
 public class RequireValidator {
+    FieldSetValidator fieldSetValidator = new FieldSetValidator();
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/federation/requiresdirective/RequireValidator.java:20]: final keyword

File: src/main/java/com/intuit/graphql/orchestrator/schema/transform/FederationTransformerPreMerge.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import static com.intuit.graphql.orchestrator.utils.XtextTypeUtils.getFieldDefinitions;
+import static com.intuit.graphql.orchestrator.utils.XtextUtils.getDirectivesWithNameFromDefinition;
+
+public class FederationTransformerPreMerge implements Transformer<XtextGraph, XtextGraph> {
+    RequireValidator requireValidator = new RequireValidator();
+    KeyDirectiveValidator keyDirectiveValidator = new KeyDirectiveValidator();
+
+    @Override
+    public XtextGraph transform(XtextGraph source) {
+        if(source.getServiceProvider().isFederationProvider()) {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/schema/transform/FederationTransformerPreMerge.java:30]: I see a common practice like below, might want to refactor:

```
if(!source.isFederationProvider())  {
  return source;
}

// do federation stuff
```

File: src/main/java/com/intuit/graphql/orchestrator/federation/ExternalValidator.java (hunk 3 above)
Comment (kmoore-intuit) [reply src/main/java/com/intuit/graphql/orchestrator/federation/ExternalValidator.java:14]: oh in the case it is not; I was practicing on creating a generic interface while doing that...I'll go ahead and remove that

**Review by CNAChino — APPROVED** [review]

**Pull request merged**